Add support for domain allow/deny list for HTTPInputSource

nishantmonu51 opened the pull request:
### Description
Right now there is no support to restrict druid HTTP based ingestion to deny ingestion from any given hosts.  
Cloud providers have metadata service that can be accessible via HTTP endpoints which can leak sensitive information about the instance on which druid is running. 
* https://docs.microsoft.com/en-us/azure/virtual-machines/linux/instance-metadata-service
* https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html

Other use-case for this config is to restrict access to internal HTTP based service endpoints that are exposed only within VPN. 

this PR adds a new config `druid.ingestion.http.allowListDomains` and  `druid.ingestion.http.denyListDomains` which allows/deny configured IP/domains from being accessed through HTTP based ingestion.

This PR has:
- [*] been self-reviewed.
- [*] added documentation for new or modified features or behaviors.
- [*] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.


<hr>

himanshug commented:
User generally is not aware of all the different domains that could leak sensitive information, so an explicit `allowList` (trying to incorporate https://searchcio.techtarget.com/feature/Rooting-out-racism-in-AI-systems-theres-no-time-to-lose ) instead-of/in-addition-to `denyList` would make the choice more secure in this context.
what do you think?

nishantmonu51 commented:
@himanshug: This config is mostly for the administrators of the druid service to secure the environment and block any internal hosts. 
The admin of the service will not have the complete allow list for the sources the service users can ingest the data from.

Although, an option for an allowList along with blacklist where the user can choose to either use allowList or denyList makes sense. I updated the PR and added both the options.

jihoonson reviewed (approved):
+1 after CI

himanshug commented on core/src/main/java/org/apache/druid/data/input/impl/HttpInputSourceConfig.java:
```diff
@@ -0,0 +1,79 @@
+      @JsonProperty("denyListDomains") List<String> denyListDomains
+  )
+  {
+    this.allowListDomains = allowListDomains == null ? Collections.emptyList() : allowListDomains;
```
it would be nice to differentiate between NULL vs EMPTY allowList ... null meaning everything is allowed and empty meaning nothing is allowed , so that a user can just set the allowList to empty to fully disable reading from anywhere.. they can add things to it as they see fit later on.

nishantmonu51 replied:
done.

himanshug commented:
+1 after https://github.com/apache/druid/pull/10535#discussion_r514609436

jihoonson commented:
Hi @nishantmonu51, I have been thinking about this PR whether it should be a part of the 0.21.0 release. My major concern is that the new security config only works for the HTTP inputSource, which means users can still use the HTTP firehose to bypass the security config. This can make a wrong impression for system administrators that they can secure their Druid cluster with the new config, which is not true. 

To avoid this problem, I think we have two options, either making both the HTTP inputSource and firehose to support the security config in the 0.21.0 release or not including them at all in the release. I would say the first option seems not eligible as it requires to add a new feature after code freeze which can potentially delay the release. So, I'd like to suggest reverting this PR. Since the problem described above will be fixed in https://github.com/apache/druid/pull/10677, we won't even have to deprecate the configs added in this PR if we go down this way. What do you think?

jihoonson commented:
@nishantmonu51 sorry for pinging you again, but do you have any opinion on this?

gianm commented:
I'm not @nishantmonu51 🙂 but IMO, it would make sense to revert this out of 0.21.0 in favor of a more comprehensive fix with a better API, like we're talking about in https://github.com/apache/druid/pull/10677.

jihoonson commented:
Thanks @gianm for chiming in. I'm going to revert this PR in both master and 0.21.0 branches.
